typescript/sdk/src/hook/read.ts

```diff
@@ -0,0 +1,284 @@
+      hooks.map(this.deriveHookConfig.bind(this)),
+    );
+
+    return {
+      type: HookType.AGGREGATION,
+      hooks: hookConfigs,
+    };
+  }
+
+  async deriveIgpConfig(address: Address): Promise<IgpHookConfig> {
+    const hook = InterchainGasPaymaster__factory.connect(
+      address,
+      this.provider,
+    );
+
+    const owner = await hook.owner();
+    const beneficiary = await hook.beneficiary();
+
+    const overhead: IgpHookConfig['overhead'] = {};
+    const oracleConfig: IgpHookConfig['oracleConfig'] = {};
+
+    let oracleKey: string | undefined;
+
+    for (const domainId of this.multiProvider.getKnownDomainIds()) {
+      const chainName = this.multiProvider.getChainName(domainId);
+
+      // if getExchangeRateAndGasPrice throws or destinationGasLimit returns 0
+      // then no gasOracle has been configured for the given domainId
+      try {
+        // this will throw if no gasOracle configured
+        const { tokenExchangeRate, gasPrice } =
+          await hook.getExchangeRateAndGasPrice(domainId);
+        // this will simply return 0 if not configured
+        const domainGasOverhead = await hook.destinationGasLimit(domainId, 0);
+
+        overhead[chainName] = domainGasOverhead.toNumber();
+        oracleConfig[chainName] = { tokenExchangeRate, gasPrice };
+
+        // we're going to assume that the owner of the first oracle we encounter
+        // is the owner of all the oracles referenced in this IgpHook
+        if (!oracleKey) {
+          const { gasOracle } = await hook.destinationGasConfigs(domainId);
+          const oracle = StorageGasOracle__factory.connect(
+            gasOracle,
+            this.provider,
+          );
+          oracleKey = await oracle.owner();
+        }
+      } catch (error) {
+        // do nothing and continue iterating through known domain IDs
+      }
+    }
+
+    return {
+      owner,
+      type: HookType.INTERCHAIN_GAS_PAYMASTER,
+      beneficiary,
+      oracleKey: oracleKey ?? owner,
+      overhead,
+      oracleConfig,
+    };
+  }
+
+  async deriveProtocolFeeConfig(
+    address: Address,
+  ): Promise<ProtocolFeeHookConfig> {
+    const hook = ProtocolFee__factory.connect(address, this.provider);
+
+    const owner = await hook.owner();
+    const maxProtocolFee = await hook.MAX_PROTOCOL_FEE();
+    const protocolFee = await hook.protocolFee();
+    const beneficiary = await hook.beneficiary();
+
+    return {
+      owner,
+      type: HookType.PROTOCOL_FEE,
+      maxProtocolFee: maxProtocolFee.toString(),
+      protocolFee: protocolFee.toString(),
+      beneficiary,
+    };
+  }
+
+  async deriveOpStackConfig(address: Address): Promise<OpStackHookConfig> {
+    const hook = OPStackHook__factory.connect(address, this.provider);
+
+    const messengerContract = await hook.l1Messenger();
+    const destinationDomain = await hook.destinationDomain();
+    const destinationChainName =
+      this.multiProvider.getChainName(destinationDomain);
```

Comment:
I worry that this will be fragile
maybe `destinationChain` should be `ChainNameOrId`